### PR TITLE
west: Add `west rtt` command

### PR DESCRIPTION
**Pull request by topisani**
Implemented in the pyocd and openocd runners while awaiting some feedback.

This command runs separately from a debug session. We should probably also support attaching a rtt client to a running debugserver. However, this might work differently for each runner. pyocd is easy, as it supports a separate RTT command, while openocd and jlink will always need a running server.

See the FIXME's for other points i would like input on.

### CODE REVIEW

**Comment by MarkoSagadin** [comment]
This would be really awesome to have. For our internal processes I created a [wrapper](https://github.com/IRNAS/irnas-east-software/blob/main/src/east/system_commands/jlink_commands.py) around JLinkExeRTTViewer so we could have a single command that does that.

How would enabling a running server be implemented in the case of JLink? As a separate command or an additional flag to the `west flash` command?

**Comment by topisani** [comment]
> How would enabling a running server be implemented in the case of JLink? As a separate command or an additional flag to the `west flash` command?

I think the two possible approaches, of which we should maybe do both as separate commands, are:
 1. `west rtt -r openocd` will run the `openocd` server just like `west debugserver` does, and attach an rtt client to it.
 2. `west rtt -r openocd` requires you to run `west debugserver` in another terminal and connects to that.
 
 The first approach is easy, when all you want is to see the log, but the second is more flexible and compatible with the other debug/debugserver/attach commands.

**Comment by MarkoSagadin** [comment]
I think that second approach makes sense, like you said, it is compatible with other commands.

Two things:

1. Is explicit `-r openocd` flag needed always? Or just in this PR? I would expect that I can just run `west rtt` regardless of the runner, if the default one support this, of course.
2. What about if the `west rtt` command checks if there is a running server (by checking known ports maybe?) and creates a new one (like suggested in your first case) if none is found? This would then support the easy use case that you describe.  So something like:
    ```
    west build -b <some_board> && west flash && west rtt
    ```

**Comment by topisani** [comment]
>     1. Is explicit `-r openocd` flag needed always? Or just in this PR? I would expect that I can just run `west rtt` regardless of the runner, if the default one support this, of course.

No, i just always use it, because the default runner is usually not what i want
 
>     2. What about if the `west rtt` command checks if there is a running server (by checking known ports maybe?) and creates a new one (like suggested in your first case) if none is found? This would then support the easy use case that you describe.  So something like:

I consideren this option, but i think its nicer to have it either as a separate command, or as an argument. Connecting to arbitrary ports when not asked to is probably a no go. Besides, consistent behavior seems preferable

File: scripts/west_commands/runners/openocd.py [hunk 1]
```diff
@@ -12,6 +12,11 @@
 from os import path
 from pathlib import Path
 from zephyr_ext_common import ZEPHYR_BASE
+import socket
+import selectors
+import sys
+import time
+import shutil
```
Comment (pdgendt) [inline scripts/west_commands/runners/openocd.py:19]: Nit: maybe put the import in alphabetical order as you're adding a bunch of new ones.

File: scripts/west_commands/runners/openocd.py [hunk 2]
```diff
@@ -388,3 +401,89 @@ def do_debugserver(self, **kwargs):
...
+            rtos_command = '${} configure -rtos Zephyr'.format(self.target_handle)
+            pre_init_cmd.append(rtos_command)
+
+        rtt_address = self.get_rtt_address()
+        if rtt_address is None:
+            raise ValueError("RTT Control block not be found")
+
+        rtt_cmds = [
+            '-c', 'rtt setup 0x{:x} 0x10 "SEGGER RTT"'.format(rtt_address),
+            '-c', 'rtt server start {} 0'.format(self.rtt_port),
```
Comment (pdgendt) [inline scripts/west_commands/runners/openocd.py:422]: Use formatted strings

```suggestion
            '-c', f'rtt setup 0x{rtt_address:x} 0x10 "SEGGER RTT"',
            '-c', f'rtt server start {self.rtt_port} 0',
```

File: scripts/west_commands/runners/openocd.py [hunk 3]
```diff
@@ -388,3 +401,89 @@ def do_debugserver(self, **kwargs):
...
+            '-c', 'rtt start',
+        ]
+
+        server_cmd = (self.openocd_cmd + self.cfg_cmd +
+                      ['-c', 'tcl_port {}'.format(self.tcl_port),
+                       '-c', 'telnet_port {}'.format(self.telnet_port),
+                       '-c', 'gdb_port {}'.format(self.gdb_port)] +
```
Comment (pdgendt) [inline scripts/west_commands/runners/openocd.py:429]: Same here

```suggestion
                      ['-c', f'tcl_port {self.tcl_port}',
                       '-c', f'telnet_port {self.telnet_port}',
                       '-c', f'gdb_port {self.gdb_port}'] +
```

File: scripts/west_commands/runners/pyocd.py [hunk 4]
```diff
@@ -213,3 +215,23 @@ def debug_debugserver(self, command, **kwargs):
...
+
+        cmd = ([self.pyocd] +
+               ['rtt'] +
+               self.pyocd_config_args +
+               self.daparg_args +
+               self.target_args +
+               self.board_args +
+               self.frequency_args +
+               self.tool_opt_args +
+               ['-a', '0x{:x}'.format(rtt_addr)])
```
Comment (pdgendt) [inline scripts/west_commands/runners/pyocd.py:235]: Same here

File: scripts/west_commands/runners/openocd.py [hunk 5]
```diff
@@ -388,3 +401,89 @@ def do_debugserver(self, **kwargs):
...
+
+    def do_rtt(self, **kwargs):
+        pre_init_cmd = []
+        for i in self.pre_init:
+            pre_init_cmd.append("-c")
+            pre_init_cmd.append(i)
+
+        if self.thread_info_enabled and self.supports_thread_info():
+            pre_init_cmd.append("-c")
+            rtos_command = '${} configure -rtos Zephyr'.format(self.target_handle)
```
Comment (pdgendt) [inline scripts/west_commands/runners/openocd.py:413]: And here

**Comment by pdgendt** [comment]
Looks very neat BTW

**Review by pdgendt — DISMISSED** [review]

**Comment by topisani** [comment]
LMK if i should have pushed this latest commit to a different PR, now that this had been partially approved. I just thought id better get this jlink support in along with the rest.

jlink support is implemented in much the same way as openocd, except it can better automatically locate the rtt block, so it relies on that.

**Review by fabiobaltieri — APPROVED** [review]
Cool, second commit message is stale I guess, feel free to update it.

**Comment by topisani** [comment]
> Cool, second commit message is stale I guess, feel free to update it.

Any objections to me just squashing all three commits?

**Comment by fabiobaltieri** [comment]
> Any objections to me just squashing all three commits?

Yes, don't. :-) Just reword the second commit message, the rest is fine as it is.

**Comment by topisani** [comment]
alright, reworded and reordered, apparently i had swapped the first two commits somewhere along the way

**Review by fabiobaltieri — APPROVED** [review]

**Review by pdgendt — APPROVED** [review]

**Comment by pdgendt** [comment]
A follow-up PR to add this to the [documentation](https://docs.zephyrproject.org/latest/develop/west/build-flash-debug.html#west-runner) would be nice.

**Pull request merged**